packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts

```diff
@@ -0,0 +1,117 @@
+  UpdatePersistedEntityMutationVariables,
+} from "../../graphql/apiTypes.gen";
+import { mustGetEntity, Subgraph } from "../../lib/subgraph";
+import { AuthenticatedUser, constructAuthenticatedUser } from "../../lib/user";
+import { updatePersistedEntityMutation } from "../../graphql/queries/knowledge/entity.queries";
+import { useAuthenticatedUser } from "./useAuthenticatedUser";
+
+type UpdateAuthenticatedUserParams = {
+  shortname?: string;
+  preferredName?: string;
+};
+
+export const useUpdateAuthenticatedUser = () => {
+  const { authenticatedUser, refetch } = useAuthenticatedUser();
+
+  const [updatePersistedEntity] = useMutation<
+    UpdatePersistedEntityMutation,
+    UpdatePersistedEntityMutationVariables
+  >(updatePersistedEntityMutation, { errorPolicy: "all" });
+
+  const [loading, setLoading] = useState<boolean>(false);
+
+  const updateAuthenticatedUser = useCallback(
+    async (
+      params: UpdateAuthenticatedUserParams,
+    ): Promise<{
+      updatedAuthenticatedUser?: AuthenticatedUser;
+      errors?: readonly GraphQLError[] | undefined;
+    }> => {
+      if (!params.shortname && !params.preferredName) {
+        return { updatedAuthenticatedUser: authenticatedUser };
+      }
+
+      const [
+        {
+          data: { me: latestSubgraph },
+        },
+      ] = await refetch();
+
+      const entityId = latestSubgraph.roots[0]!;
+
+      /** @todo: use a partial update mutation instead */
+      const { properties: currentProperties } = mustGetEntity({
+        subgraph: latestSubgraph as unknown as Subgraph,
+        entityId,
+      });
+
+      const { errors } = await updatePersistedEntity({
+        variables: {
+          entityId,
+          updatedProperties: {
+            ...currentProperties,
+            ...(params.shortname
+              ? {
+                  [extractBaseUri(types.propertyType.shortName.propertyTypeId)]:
+                    params.shortname,
+                }
+              : {}),
+            ...(params.preferredName
+              ? {
+                  [extractBaseUri(
+                    types.propertyType.preferredName.propertyTypeId,
+                  )]: params.preferredName,
+                }
+              : {}),
+          },
+        },
+      });
+
+      if (errors && errors.length > 0) {
+        return { errors };
+      }
+
+      const [
+        {
+          data: { me: updatedSubgraph },
+        },
+        kratosSession,
+      ] = await refetch();
+
+      const updatedAuthenticatedUser = constructAuthenticatedUser({
+        userEntityId: updatedSubgraph.roots[0]!,
+        /**
+         * @todo: ensure this subgraph contains the incoming links of orgs
+         * at depth 2 to support constructing the `members` of an `Org`.
+         *
+         * @see https://app.asana.com/0/1202805690238892/1203250435416412/f
+         */
+        subgraph: updatedSubgraph as unknown as Subgraph,
+        kratosSession: kratosSession!,
```

Comment:
Is this safe?

Reply:
I think so, yes - because if `kratosSession` is `undefined` an error would've been thrown in the above queries (as they require the user to be authenticated with a `kratosSession`. Would you like me to add a comment briefly explaining this?

Reply:
Does the variable even need to be optional then? Just wondering if we can catch it earlier and not need the assertion

Reply:
I can't think of a clean solution, as `kratosSession` has to be `undefined` at first render because it is fetched asynchronously. So in this function I can throw if its `undefined`, return if its `undefined`, but that would be dead code because as I said the queries will fail first.

Reply:
Have added the explicit check in https://github.com/hashintel/hash/pull/1299/commits/320831c20bbbefca558192427ea2d6d0fe4f2ed8.

> if we refactored the code here, re-ordered things, or changed to queries that don't require it, and TSC wouldn't warn us of that.

I'm not sure how this explicit check would let TSC warn us about anything in these scenarios, but I don't feel strongly and probably nice to have it be more explicit anyway.